### ISSUE TITLE
DataDesc concerns

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1835]
While working on #1755 there were some things I noticed about our data description language, `DataDesc`, that seemed like they could be issues, but I'm not sure so I'm making this issue for feedback. For context, here's the definition of the language:
https://github.com/JacquesCarette/Drasil/blob/f86a03034d3c1f9d0a40b9019e8e61c13f61107e/code/drasil-code/Language/Drasil/Code/DataDesc.hs#L8-L20

#### 1. Some data constructors only make sense with certain types of data

Specifically:
- `Line (Repeat d) _` describes a line containing data for `CodeChunk`s `d` in a repeating pattern. This only makes sense for `d` with list-types, because non-list data would just be overwritten with each repeat of the pattern
- `Lines (Straight d) _ _` describes data for `CodeChunk`s `d` spread across multiple lines, where each line does not have a repeating pattern. This also only makes sense for `d` with list-types, for the same reason as above.
- `Lines (Repeat d) _ _` describes data for `CodeChunk`s `d` spread across multiple lines where each line does have a repeating pattern. This only makes sense for 2D list types, because there are essentially 2 levels of repetition (the explicit `Repeat` line pattern and the inter-line repetition pattern implicit in the use of `Lines`)

The potential problem here is that creating a `DataDesc` with a non-list type `CodeChunk` in a place that only makes sense for list-types is still considered valid. You would probably hit an error when using the `DataDesc` in a function that checks the types, but it feels like the error should happen earlier, or it should be impossible to create such a `DataDesc` in the first place.

Also, in writing this out, it occurs to me that `DataDesc` implicitly assumes the newline as a delimiter. `Lines (Straight [d]) _ _` is equivalent to `Line (Repeat [d]) '\n'`. It feels like the language should allow defining delimiters for multiple levels, rather than basing the data description around line breaks.

#### 2. Lists cannot be on the same line as non-list data

Say `x` and `z` are scalar and `y` is a vector and the desired input format is to have all values on one line, something like `x, y_1, y_2, z`. This can be described in `DataDesc` by:  `Line (Straight [x,y,z]) ','`. But this description does not give enough information to be useful. Anything trying to read such data would need to know the size of `y` ahead of time so it knows when it has switched to reading the value for `z`.

If we want to support this kind of scenario, it again seems like we would need to be able to describe a delimiter for a second level. If we could say that `;` is the outermost delimiter and `,` is the delimiter for the next level inward, then we could read lists and non-lists on the same line with a format like `x; y_1, y_2; z`

#### 3. `Lines _ Nothing _` only makes sense as the last item

 `Lines` accepts a `Maybe Integer` for the number of lines. If unknown, `Nothing` means to read until the end of the file. So that means it doesn't make sense to have a `DataDesc` with more `DataItem`s after a `Lines _ Nothing _` item, unless the subsequent items are all `JunkData`. But, similar to point 1 above, a `DataDesc` with more items after a `Lines _ Nothing _` would still be considered valid.

#### Conclusion

After typing these out, it does seem like they boil down to 2 potential problems:
- There are valid `DataDesc`s that don't actually make sense
- The lack of ability to specify multiple levels of delimiter forbids certain formats

The first might be fine, as we can always check for true validity and throw errors if something doesn't make sense, though the second seems to me like it could be a real problem with `DataDesc`. Thoughts @JacquesCarette or anyone else?

**Comment by JacquesCarette** [comment]
Yes, `DataDesc` was a first stab at a data description language. And you have noticed that it is 'untyped', so that things can indeed go wrong.

We were not going for ultimate flexibility (thus the lack of ability to specify certain things). And because it was an early experiment, not types.

The idea, as always with entirely new features of Drasil, was to put something partly reasonable in place [but inspired from existing solutions, so that we don't stray too far into obviously-wrong territory], see what the actual requirements are, then refactor as needed.

Basically what we need is to understand what kinds of file formats we're likely to want to read/write, then implement just enough to cover that. We'd prefer to not box ourselves in too much, so that we should plan to be able to expand it later. And it would be nice to have it be typed, so that silly specifications can be ruled out -- within reason.

There is nothing set in stone about the current `DataDesc`. If you can come up with a better design, go for it!

**Comment by JacquesCarette** [comment]
Nice idea. We can 'type check' the `DataDesc` before we use it, so if it does bad things, we can give an error message early.  Not quite at compile time, but early enough.

**Comment by bmaclach** [comment]
We also want to be able to read binary data, as mentioned in https://github.com/JacquesCarette/Drasil/issues/1716#issuecomment-527555167, so we should also tag `Data` with one more piece of information, something like `data Encoding = Binary | Text`.

**Comment by JacquesCarette** [comment]
@bmaclach you did implement some of this, right? Is this still a live issue?

**Comment by bmaclach** [comment]
I tried to implement the new `DataDesc` design but it proved to be more difficult than it seemed. My work is on the `newDataDesc` branch. First off, I had to change the design slightly from what was proposed above. The "current" new design is:
https://github.com/JacquesCarette/Drasil/blob/01dc788f4ecbfa4cd87dc85c0bdc9c2d45bff320/code/drasil-code/Language/Drasil/Code/DataDesc.hs#L12-L41

Currently we use `DataDesc` in 3 applications:
1. For reading the sample data file (`readWithDataDesc` function)
2. For writing the sample input file (`convDataDesc` function)
3. For generating code that reads files (`readData` function)

I updated 1 in the `newDataDesc` branch to work with the new `DataDesc'` design. It works, but implementing it was really messy. `readWithDataDesc` pattern-matches on the `DataDesc` AST, and I couldn't write a generic pattern-match for the `Data` node. Instead, I had to also match on the dimension of the data and the degree of intermixing. One of the strengths of the new design is that it can describe n-dimensional data, whereas the old `DataDesc` could only describe up to 2-dimensional data, but due to being unable to write a pattern-match case that was generic on the dimension, I only wrote cases for 1-dimensional and 2-dimensional data, then stopped.

Now, the reasons I couldn't write a generic pattern-match case:
- The only `Expr` node for lists is `Matrix :: [[Expr]] -> Expr`, which "assumes" 2 dimensions. You can get more or less than 2 dimensions by nesting, but the fact that `Matrix` is parameterized by a 2D list makes the logic that would be required in `readWithDataDesc` really, really complicated, and I couldn't wrap my head around it. It would be much nicer if we had a 1D list node in Expr, say `Vect`.
- If we had the 1D `Vect` node, we would also need an `Expr` level `map` function to map a function over a `Vect` `Expr`, and an `Expr` level `transpose` function to transpose a `Vect` `Expr`.

I wrote what the generic pattern-match case would look like as comments in the code, so that once we have the things I mentioned above, we should be able to uncomment that and have a great `readWithDataDesc` implementation.

But the other problem I ran into was that GOOL's language for input-reading, like the current `DataDesc`, assumes that newlines will be a delimiter when reading data. So when it comes to updating number 3 above, the `readData` function, GOOL's available functions are not enough to handle the range of file formats that the new `DataDesc` design can describe. What we need is a better language for generating I/O statements using GOOL. We have an issue open related to this: #1958. The issue talks about "printing combinators", but the solution should really work in both directions -- reading input as well as printing output. In fact, the new `DataDesc` design should maybe be defined directly in GOOL, and have GOOL's I/O functions be parameterized by `DataDesc`, to solve both issues at once.

Since there was so much to do for this and so little time, and it didn't seem like a high priority, I stopped working on this and moved on to writing my thesis. 

I suppose I could actually create a PR from the `newDataDesc` branch, since everything there works. The state of things on that branch is: we still have the old `DataDesc`, but also the new design, `DataDesc'`. `readWithDataDesc` uses the new design, but everything else still uses the old design. If we get it merged to master, then my work will be there for the next person to work on this (I left a lot of comments too so that they hopefully can understand what I was doing). But I don't know how you would feel about having both versions of `DataDesc` live in master at the same time, @JacquesCarette?

**Comment by JacquesCarette** [comment]
Very thorough description, thank you. That helps a lot.  The need for synchronization between the various pieces of Drasil is fascinating!  And how the embedded assumptions in that code were road blocks is too.  Hmm, something for a 'future work' in your thesis? Basically cut-and-paste parts of the above?

Good question about the mixed design on `master`.  If we leave your work on a branch, it will just die from code-rot. If we bring it in, we have some weird hybrid design, but still better than what was there before.

So I'm going to say: merge.

**Comment by smiths** [comment]
My vote doesn't have the same weight on code related issues, but I'm glad that @JacquesCarette suggested merging in the code by @bmaclach.  I was thinking the same thing.  If we don't merge, we will inevitably lose the work, and it does sound like it is an improvement in the right direction.
